File: January2021/GetMaximumInGeneratedArray.java
class Solution {

    //TC : O(n)
    // SC: O(n)
    public int getMaximumGenerated(int n) {
        int[] array = new int[n+1];
        array[0] = 0;
        if(n==0){
            return 0;
        }
        array[1] = 1;

        int max = 1;
        for(int i=2;i<=n;i++){
            if(i%2==0){
                array[i] = array[i/2];
            } else{
                array[i] = array[i/2] + array[i/2+1];
            }
            max = Math.max(max, array[i]);
        }
        return max;
    }
}

File: January2021/C++/GetMaximuminGeneratedArray.cpp
/**
 * @author ssrbazpur
 * Time Complexity:O(N)
 * Space Complexity:O(N)
 *
**/

class Solution
{
public:
    int getMaximumGenerated(int n)
    {
        vector<int> nums(n + 1, 0);

        int ans = 0;

        if (n >= 1)
        {
            ans = 1;
            nums[1] = 1;
        }

        for (int i = 1; 2 * i + 1 <= n; i++)
        {
            nums[2 * i] = nums[i];
            nums[2 * i + 1] = nums[i] + nums[i + 1];
        }

        for (int n : nums)
        {
            ans = max(ans, n);
        }

        return ans;
    }
};